src/fparser/two/Fortran2003.py

```diff
@@ -7203,102 +7208,472 @@ def match(string):
+                    current_string = repmap(splitted[1]).strip()
+        if len(item_list) <= 1:
+            # a list must contain at least 2 items (see SequenceBase)
+            return None
+        return ',', tuple(item_list)
+
+
+class Format_Specification(BracketBase):  # pylint: disable=invalid-name
+    '''
+    Fortran 2003 rule R1002
+
+    format-specification = ( [ format-item-list ] )
+
+    C1002 is implemented in a separate class Format_Item_C1002
+
+    C1002 (R1002) The comma used to separate format-items in a
+    format-item-list may be omitted
+
+    (1) Between a P edit descriptor and an immediately following F, E,
+    EN, ES, D, or G edit descriptor, possibly preceded by a repeat
+    specifier,
+
+    (2) Before a slash edit descriptor when the optional repeat
+    specification is not present,
+
+    (3) After a slash edit descriptor, or
+
+    (4) Before or after a colon edit descriptor.
+
+    '''
     subclass_names = []
     use_names = ['Format_Item_List']
 
     @staticmethod
     def match(string):
+        '''Implements the matching for a format specification.
+
+        :param str string: The string to check for conformance with a \
+                           format specification.
+        :return: `None` if there is no match, otherwise a tuple of \
+        size three, the first entry being a string containing a left \
+        bracket and the third being a string containing a right \
+        bracket. The second entry is either a Format_Item or a \
+        Format_Item_List.
+        :rtype: `NoneType` or ( `str`, \
+        :py:class:`fparser.two.Fortran2003.Format_Item` or \
+        :py:class:`fparser.two.Fortran2003.Format_Item_List`, `str` )
+
+        '''
         return BracketBase.match('()', Format_Item_List, string,
                                  require_cls=False)
 
+def skip_digits(string):
+    '''Skips over any potential digits (including spaces) to the next
+    non-digit character and return its index. If no such character is
+    found or if the first character in the string is not a digit then
+    specify that the skip has failed.
 
-class Format_Item_C1002(Base):  # C1002
-    """
-::
-    <format-item-c1002> = <k>P [,] (F|D)<w>.<d> | (E|EN|ES|G)<w>.<d>[E<e>]
-                          | [<r>]/ [,] <format-item>
-                          | : [,] <format-item>
-                          | <format-item> [,] / [[,] <format-item>]
-                          | <format-item> [,] : [[,] <format-item>]
+    :param str string: The string to search
+    :returns: a 2-tuple with the first entry indicating if a valid \
+    character has been found and the second entry indicating the index \
+    of this character in the 'string' argument.
+    :rtype: (bool, int or NoneType)
 
-Attributes
-----------
-items : (Format_Item, Format_Item)
-    """
+    '''
+    found = False
+    index = None
+    if string[0].isdigit():
+        index = 0
+        while True:
```

Comment:
Sorry to be picky but wouldn't it be nicer to use `enumerate` here:

            for index, char in enumerate(string):
                if not (char.isdigit() or char == ' '):
                    found = True
                    break

Reply:
I don't think you are being picky. It is much nicer your suggested way. I still needed to add a special case for the first index.

Reply:
Regarding putting skip_chars in utils, my approach has been that if something is used in more than one file then it should go in utils. This function is quite specialised and is only used in Fortran2003.py so my view was (and is) that it should stay where it is.

Reply:
Fine.